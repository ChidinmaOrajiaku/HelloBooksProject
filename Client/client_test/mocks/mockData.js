import { updatePassword } from "../../actions/updatePassword";

const mockData = {
  userDetails: {
    message: 'Welcome admin',
    username: 'admin',
    token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImRhZGR5Iiwicm9sZSI6InVzZXIiLCJpZCI6MiwiaWF0IjoxNTEzODc1ODE3LCJleHAiOjE1MTM5NjIyMTd9.oexUFhi08vhMU88LmZJvRDy9JFdThMpfb3Qz9DWWzgk'
  },
  userDetailsResponse: {
    username: 'daddy',
    role: 'user',
    id: 2,
    iat: 1513875817,
    exp: 1513962217
  },
  userCountResponse: {
    count: 7
  },
  borrowBookData: {
    bookId: 1,
    userId: 1,
  },
  borrowBookDataResponse: {
    returned: false,
    id: 24,
    userId: 1,
    bookId: 7,
    toReturnDate: '2018-01-14',
    updatedAt: '2017-12-21T21:16:08.442Z',
    createdAt: '2017-12-21T21:16:08.442Z',
    returnDate: null,
    categoryId: null
  },
  createBookData: {
    title: 'a',
    author: 'a',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  createBookErrorData: {
    author: 'a',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  createBookErrorResponse: 'An error occured',
  createBookResponse: {
    message: 'Succesfully added'
  },
  getABookData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getAllBooksData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getAllBorrowedBooksData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getCategoryData: {
    category: 'Fiction'
  },
  getUserData: {
    firstname: 'daddy',
    lastname: 'daddy',
    username: 'daddy',
    password: 'daddy',
    email: 'daddy@gmail.com'
  },
  bookData: {
    title: 'a',
    author: 'b',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  returnBookData: {
    message: 'Successfully Returned'
  },
  userData: {
    verifyPassword: '1234',
    password: '4321'
  },
  errorMessage: 'An error occurred',
  image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1515403427/ena7dhy6xriztom6vuvo.jpg'
};

export default mockData;
