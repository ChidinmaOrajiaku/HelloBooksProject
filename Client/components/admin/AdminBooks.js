import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { getRequest } from '../../actions/getAllBooks';
import { adminModifyRequest } from '../../actions/modifyBooks';
import { adminDeleteRequest } from '../../actions/deleteBooks';
import { admingetBorrowedRequest } from '../../actions/getAllBorrowedBooks';


/**
 * @class Books
 */
class Books extends React.Component {
  /**
     * @constructor
     * @param {object} props 
     */
  constructor(props) {
    super(props);
    this.state = {
      getAllBooks: [],
      getAllBorrowedBooks: [],
      bookStatus: '',
      loading: true,
      pointer: false,
    };
  }
  /**
 * 
 * @returns {event} handles change
 * @param {any} event 
 * @memberof Books
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
   * 
   * @returns {nextProps} next props
   * @param {any} nextProps 
   * @memberof Books
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      loading: false,
      getAllBooks: nextProps.getAllBooksData,
      getAllBorrowedBooks: nextProps.getAllBorrowedBooksData
    });
  }
  /**
   * 
   * @returns {props} actions
   * @memberof Books
   */
  componentDidMount() {
    this.props.getRequest();
    this.props.admingetBorrowedRequest();
    $('.tooltipped').tooltip({ delay: 50 });
    $('select').material_select();
    $('select').change(event => this.handleChange(event));
  }
  /**
     * 
     * 
     * @constructor 
     * @memberof Books
     */
  render() {
    const books = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="teal-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.getAllBooks).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.getAllBooks[key].image }/> </td>
              <td>{ this.state.getAllBooks[key].title }</td>
              <td>{ this.state.getAllBooks[key].author }</td>
              <td>{ this.state.getAllBooks[key].category }</td>
              <td><i className="material-icons">zoom_in</i><i className="material-icons">create</i><i className="material-icons">delete</i></td>
            </tr>
          )}
        </tbody>
      </table>;
    const borrowedBooks = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="teal-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.getAllBorrowedBooks).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.getAllBorrowedBooks[key].Book.image }/> </td>
              <td>{ this.state.getAllBorrowedBooks[key].Book.title }</td>
              <td>{ this.state.getAllBorrowedBooks[key].Book.author }</td>
              <td>{ this.state.getAllBorrowedBooks[key].Book.category }</td>
            </tr>
          )}
        </tbody>
      </table>;
    return (
      <div className="books row ontainer">
        <div className=""> <NavigationBar /> </div>
        <div className="row col m8 offset-m3">
          <div className="input-field col s4">
            {<select className="teal-text" id= "bookStatus"value="1" onChange={this.handleChange}>
              <option value="" defaultValue>Choose your option</option>
              <option value="2">All Books</option>
              <option value="3">Pending Returns</option>
            </select>}
            <label className="teal-text">Select View</label>
          </div>
          <div className="col s12 ">
            <div className="card">
              <div className="card-content white-text">
                {this.state.bookStatus === '3' ? borrowedBooks : books }
              </div>
            </div>
          </div>
        </div>
        <div> <Footer/></div>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    getAllBooksData: state.getAllBooks[0].response,
    getAllBorrowedBooksData: state.getAllBorrowedBooks[0].response
  }
);

export default connect(mapStateToProps,
  { adminModifyRequest, adminDeleteRequest, admingetBorrowedRequest, getRequest })(Books);
