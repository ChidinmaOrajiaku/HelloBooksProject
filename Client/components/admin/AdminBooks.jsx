import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { getRequest } from '../../actions/getAllBooks';
import { adminDeleteRequest } from '../../actions/deleteBooks';
import { admingetBorrowedRequest } from '../../actions/getAllBorrowedBooks';
import { getBookRequest } from '../../actions/getABook';
import { editBookIdRequest } from '../../actions/editBooks';


/**
 * @class Books
 */
class AdminBooks extends React.Component {
  /**
     * @constructor
     * @param {object} props 
     */
  constructor(props) {
    super(props);
    this.state = {
      bookId: '',
      bookIndex: '',
      deleteBook: '',
      filterData: [],
      getABook: [],
      getAllBooks: [],
      getAllBorrowedBooks: [],
      bookStatus: '',
      loading: true,
      pointer: false,
    };
    this.onViewRequest = this.onViewRequest.bind(this);
    this.onDeleteRequest = this.onDeleteRequest.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleEditChange = this.handleEditChange.bind(this);
    this.handleDeleteChange = this.handleDeleteChange.bind(this);
  }

  /**
   * 
   * @returns {object} response object
   * @memberof AdminBooks
   */
  componentDidMount() {
    this.props.getRequest();
    this.props.admingetBorrowedRequest();
    $('select').material_select();
    $('select').change(event => this.handleChange(event));
    $('.modal').modal({
      dismissible: true,
      opacity: 0.5,
      inDuration: 300,
      outDuration: 200,
      startingTop: '2%',
      endingTop: '20%',
    });
  }
  /**
   * 
   * @returns {nextProps} next props
   * @param {any} nextProps 
   * @memberof AdminBooks
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      loading: false,
      getAllBooks: nextProps.getAllBooksData,
      filterData: nextProps.getAllBooksData,
      getAllBorrowedBooks: nextProps.getAllBorrowedBooksData,
      getABook: nextProps.getABookData,
      deleteBook: nextProps.deleteBookData,
      modifyBook: nextProps.modifyBookData
    });
  }

  /**
 * 
 * @returns {event} handles change
 * @param {any} event 
 * @memberof AdminBooks
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
 * 
 * @returns {event} handles delete change
 * @param {any} event 
 * @memberof AdminBooks
 */
  handleDeleteChange(event) {
    this.setState({
      bookId: event.target.value,
      bookIndex: event.target.dataset.index });
  }

  /**
 * 
 * @returns {event} handles edit change
 * @param {any} event 
 * @memberof AdminBooks
 */
  handleEditChange(event) {
    event.persist();
    this.props.editBookIdRequest(event.target.value);
    localStorage.setItem('currentBookId', event.target.value);
  }
  /**
 * 
 * @returns {event} view a book request
 * @param {any} event 
 * @memberof AdminBooks
 */
  onViewRequest(event) {
    event.preventDefault();
    this.props.getBookRequest(event.target.value);
  }
  /**
 * 
 * @returns {event} delete a book request
 * @param {any} event 
 * @memberof AdminBooks
 */
  onDeleteRequest(event) {
    event.preventDefault();
    this.props.adminDeleteRequest(this.state.bookId);
    setTimeout(() => {
      if (this.props.deleteBookData.isDeleted === true) {
        this.state.filterData.splice(this.state.bookIndex, 1);
        this.setState({
          getAllBooks: this.state.filterData
        });
        this.props.history.push('/books');
        Materialize.toast('Successfully deleted', 2000, 'teal rounded');
      } else {
        Materialize.toast('Not Deleted', 2000, 'red rounded');
      }
    }, 1000);
  }
  /**
     * 
     * 
     * @returns {object} ReactMarkupElement 
     * @memberof AdminBooks
     */
  render() {
    const books = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="white-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.getAllBooks).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.getAllBooks[key].image }/> </td>
              <td>{ this.state.getAllBooks[key].title }</td>
              <td>{ this.state.getAllBooks[key].author }</td>
              <td>{ this.state.getAllBooks[key].category }</td>
              <td>
                <Link to="/editbook">
                  <button
                    value={this.state.getAllBooks[key].id}
                    onClick={this.handleEditChange}
                    data-index= {key}
                    className="material-icons">create
                  </button>
                </Link>
                <a href="#modal1" className="modal-trigger">
                  <button
                    value={this.state.getAllBooks[key].id}
                    onClick={this.handleDeleteChange}
                    data-index= {key}
                    className="material-icons">delete
                  </button>
                </a>
              </td>
            </tr>
          )}
        </tbody>
      </table>;
    const borrowedBooks = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="white-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.getAllBorrowedBooks).map(key =>
            <tr key={key}>
              <td>
                <img src = {this.state.getAllBorrowedBooks[key].Book.image }/>
              </td>
              <td>{ this.state.getAllBorrowedBooks[key].Book.title }</td>
              <td>{ this.state.getAllBorrowedBooks[key].Book.author }</td>
              <td>{ this.state.getAllBorrowedBooks[key].Book.category }</td>
            </tr>
          )}
        </tbody>
      </table>;

    return (
      <div className="books row container-fluid">
        <div> <NavigationBar /> </div>
        <h4 className="col m8 offset-m3 white-text"> ADMIN BOOK LIST </h4>
        <div className="row col m8 offset-m3">
          <div className="input-field col s4 status">
            <select
              className="white-text"
              id= "bookStatus"
              value="1"
              onChange={this.handleChange}>
              <option value="" defaultValue>Choose your option</option>
              <option value="2">All Books</option>
              <option value="3">Pending Returns</option>
            </select>
            <label className="white-text sort">Sort table</label>
          </div>
          <div className="col s12 ">
            <div className="card">
              <div className="card-content teal-text">
                {this.state.bookStatus === '3' ? borrowedBooks : books }
              </div>
            </div>
          </div>
        </div>
        <div id="modal1" className="modal">
          <div className="modal-content">
            <p className="white-text modalDelete">
              Are you you want to delete this book?
            </p>
          </div>
          <div className="modal-footer">
            <button
              onClick={this.onDeleteRequest}
              className="deleteButton"> Delete
            </button>
            <a className="modal-action modal-close">
              <button className="cancelButton">Cancel
              </button>
            </a>
          </div>
        </div>
        <div> <Footer /></div>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    getAllBooksData: state.getAllBooks[0].response,
    getAllBorrowedBooksData: state.getAllBorrowedBooks[0].response,
    getABookData: state.getABook[0].response,
    deleteBookData: state.deleteBooks[0]
  }
);

export default connect(mapStateToProps,
  { adminDeleteRequest, admingetBorrowedRequest, getRequest, getBookRequest, editBookIdRequest
  })(AdminBooks);
