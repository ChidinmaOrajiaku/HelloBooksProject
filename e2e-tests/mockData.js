module.exports = {
  file: {
    lastModified: 1508779018000,
    lastModifiedDate: 'Mon Oct 23 2017 18:16:58 GMT+0100 (WAT)',
    name: '3 card game.jpg',
    size: 642971,
    type: 'image/jpeg',
    webkitRelativePath: ''
  },
};

