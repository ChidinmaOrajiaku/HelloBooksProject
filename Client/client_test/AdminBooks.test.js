import React from 'react';
import $ from 'jquery';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import mockData from './mocks/mockData';
import { AdminBooks, mapStateToProps } from '../components/admin/AdminBooks.jsx';

configure({ adapter: new Adapter() });

describe('<AdminBooks />', () => {
  const props = {
    getUserBorrowed: jest.fn(),
    yetToReturn: jest.fn(),
    getBookRequest: jest.fn(),
    adminDeleteRequest: jest.fn(),
    getRequest: jest.fn(),
    admingetBorrowedRequest: jest.fn(),
    editBookIdRequest: jest.fn()
  };
  let mountedComponent;
  const AdminBooksItem = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<AdminBooks {...props}/>);
    }
    return mountedComponent;
  };

  it('should call handleDeleteChange method', () => {
    const spy = jest.spyOn(AdminBooksItem().instance(), 'handleDeleteChange');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'abd',
        dataset: {
          index: 1
        }
      }
    };
    AdminBooksItem().instance().handleDeleteChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call handleEditChange method', () => {
    const spy = jest.spyOn(AdminBooksItem().instance(), 'handleEditChange');
    const event = {
      persist: jest.fn(),
      target: {
        value: 'chideberecom'
      }
    };
    AdminBooksItem().instance().handleEditChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call onViewRequest method', () => {
    const spy = jest.spyOn(AdminBooksItem().instance(), 'onViewRequest');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'chideberecom'
      }
    };
    AdminBooksItem().instance().onViewRequest(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call onDeleteRequest method', () => {
    const spy = jest.spyOn(AdminBooksItem().instance(), 'onDeleteRequest');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'chideberecom'
      }
    };
    AdminBooksItem().instance().onDeleteRequest(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call handleChange method', () => {
    const spy = jest.spyOn(AdminBooksItem().instance(), 'handleChange');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'chideberecom'
      }
    };
    AdminBooksItem().instance().handleChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call componentWillReceiveProps method', () => {
    const nextProps = {
      getAllBooksData: [],
      getABookData: [],
      getAllBorrowedBooksData: [],
      deleteBookData: [],
      modifyBookData: [],
    };
    const spy = jest.spyOn(AdminBooks.prototype, 'componentWillReceiveProps');
    shallow(<AdminBooks {...props} componentWillReceiveProps={spy}/>)
      .instance().componentWillReceiveProps(nextProps);
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('ensures that mapStateToProps dispatches the specified actions', () => {
    const state = {
      getAllBooks: ['1', '2'],
      getABook: ['1', '2'],
      getAllBorrowedBooks: ['1', '2'],
      deleteBooks: ['1', '2'],
    };
    expect(mapStateToProps(state).getAllBooksData).toExist;
    expect(mapStateToProps(state).getABookData).toExist;
    expect(mapStateToProps(state).getAllBorrowedBooksData).toExist;
    expect(mapStateToProps(state).deleteBookData).toExist;
  });

  it('should match snapshot test', () => {
    const component = AdminBooksItem();
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
