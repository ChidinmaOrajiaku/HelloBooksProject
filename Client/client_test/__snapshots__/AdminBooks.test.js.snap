// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminBooks /> should match snapshot test 1`] = `
<div
  className="books row container-fluid"
>
  <div>
     
    <Connect(withRouter(NavigationBar)) />
     
  </div>
  <h4
    className="col m8 offset-m3 white-text"
  >
     ADMIN BOOK LIST 
  </h4>
  <div
    className="row col m8 offset-m3"
  >
    <div
      className="input-field col s4 status"
    >
      <select
        className="white-text"
        id="bookStatus"
        onChange={[Function]}
        value="1"
      >
        <option
          defaultValue={true}
          value=""
        >
          Choose your option
        </option>
        <option
          value="2"
        >
          All Books
        </option>
        <option
          value="3"
        >
          Pending Returns
        </option>
      </select>
      <label
        className="white-text sort"
      >
        Sort table
      </label>
    </div>
    <div
      className="col s12 "
    >
      <div
        className="card"
      >
        <div
          className="card-content teal-text"
        >
          <div>
            <p>
              Loading...
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="modal"
    id="modal1"
  >
    <div
      className="modal-content"
    >
      <p
        className="white-text modalDelete"
        id="modalDelete"
      >
        Are you you want to delete this book?
      </p>
    </div>
    <div
      className="modal-footer"
    >
      <button
        className="deleteButton"
        id="deleteButton"
        onClick={[Function]}
      >
         Delete
      </button>
      <a
        className="modal-action modal-close"
      >
        <button
          className="cancelButton"
        >
          Cancel
        </button>
      </a>
    </div>
  </div>
  <div>
     
    <Component />
  </div>
</div>
`;
